airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/write/StreamLoader.kt

```diff
@@ -12,32 +12,52 @@ import io.airbyte.cdk.load.message.SimpleBatch
- * [Batch] that the framework may pass to [processBatch] and/or [finalize]. (See @[Batch] for more
- * details.)
+ * [processRecords] is called whenever a batch of records is available for processing (of the size
+ * configured in [io.airbyte.cdk.load.command.DestinationConfiguration.recordBatchSizeBytes]) and
+ * only after [start] has returned successfully. The return value is a client-defined implementation
+ * of @ [Batch] that the framework may pass to [processBatch]. (See @[Batch] for more details.)
  *
- * [processBatch] is called once per incomplete batch returned by either [processRecords] or
- * [processBatch] itself.
+ * [processRecords] may be called concurrently by multiple workers, so it should be thread-safe if
+ * [io.airbyte.cdk.load.command.DestinationConfiguration.numProcessRecordsWorkers] > 1. For a
+ * non-thread-safe alternative, use [createBatchAccumulator].
+ *
+ * [createBatchAccumulator] returns an optional new instance of a [BatchAccumulator] to use for
+ * record processing instead of this stream loader. By default, it returns a reference to the stream
+ * loader itself. Use this interface if you want each record processing worker to use a separate
+ * instance (with its own state, etc).
  *
- * [finalize] is called once after all records and batches have been processed successfully.
+ * [processBatch] is called once per incomplete batch returned by either [processRecords] or
+ * [processBatch] itself. It must be thread-safe if
+ * [io.airbyte.cdk.load.command.DestinationConfiguration.numProcessBatchWorkers] > 1. If
+ * [processRecords] never returns a non-[Batch.State.COMPLETE] batch, [processBatch] will never be
+ * called.
  *
- * [close] is called once after all records have been processed, regardless of success or failure.
- * If there are failed batches, they are passed in as an argument.
+ * [close] is called once after all records have been processed, regardless of success or failure,
+ * but only if [start] returned successfully. If any exception was thrown during processing, it is
+ * passed as an argument to [close].
  */
-interface StreamLoader {
+interface StreamLoader : BatchAccumulator {
     val stream: DestinationStream
 
     suspend fun start() {}
-    suspend fun processRecords(records: Iterator<DestinationRecord>, totalSizeBytes: Long): Batch
+    suspend fun createBatchAccumulator(): BatchAccumulator = this
+
     suspend fun processFile(file: DestinationFile): Batch
     suspend fun processBatch(batch: Batch): Batch = SimpleBatch(Batch.State.COMPLETE)
     suspend fun close(streamFailure: StreamIncompleteResult? = null) {}
 }
+
+interface BatchAccumulator {
+    suspend fun processRecords(
+        records: Iterator<DestinationRecord>,
+        totalSizeBytes: Long,
+        endOfStream: Boolean = false
+    ): Batch =
+        throw NotImplementedError(
+            "processRecords must be implemented if createBatchAccumulator is overridden"
+        )
+}
```

Comment:
I feel like trying to unify the two code variants (record -> upload vs record -> part -> upload) is making this interface complex. What if we just have a separate hook for the direct record -> upload path (i.e. iceberg)?

Reply:
I think the right final form is a single hook that returns a list of accumulators, whose return types chain (plus configuration for number of workers + a flag for whether it does network IO).

Something like

`createBatchProcessingPipeline(
  BatchPipelineStep(RecordToPartAccumulator, 2, io=false),
  BatchPipelineStep(PartToObjectAccumulator, 10, io=true),
  BatchPipelineStep(ObjectToRawTableAccumulator, 5, io=true),
  etc.
`

Reply:
I agree if we want to support that level of modularity a composable pipeline is a good way to go. I quite like those interfaces. However, for now, the current interface is a bit convoluted—it took me a while to understand this. I'd suggest we have a DirectUploader and a PartUploader interface in the interim. We can then abstract as prudent going forward. I also think using `Batch` to refer to both a `Part` and a merged `Upload` to be more trouble than it's worth.

Reply:
Ok, so what you're suggesting is that we have explicit hooks for these things right now the same way we do for `processFile`? I guess that makes sense. Like, to test it with the BigQuery example: `PartUploader` would just send the prepared data straight to the DB.

Tho we did decide we'd try to avoid interface changes in the refactor. Like, the only interface change I introduced here was the bare minimum to handle the issues introduced by having a fixed number of workers. I'd have to do more work/testing here to get the right methods conditionally exercised.

Maybe we should talk through the steps to get where we want to be and do it as a separate item, to minimize the work required?